### ISSUE TITLE
Can't connect to PostgreSQL via SSL

### ISSUE THREAD

**Issue by msluser** [serge-rider/dbeaver#1835]
I am trying to connect to a PostgreSQL database server using SSL but I get the following error:

![image](https://user-images.githubusercontent.com/20219053/27767071-35114464-5ea6-11e7-838f-e938f858e39a.png)

The PostgreSQL database server is 9.6 running on Ubuntu 16.04 LTS
I am able to connect to this database using PGAdmin4 using SSL with no problems.

I used a self-signed certificate authority (CA)

Here is how I generated  the required files
ca-priv.key - my CA private key file
ca-cert.crt - my CA certificate file
server-priv-key - my server private key file
server-cert-req.csr - my server certificate request file
server-cert.crt - my server certificate file
server-cert-fullchain.crt - my server certificate that includes the full chain of certificates up to my root CA certificate
johnsmith-priv.key my user private key file
johnsmith-cert.crt my user certificate file
johnsmith-cert-req.csr my user certificate request file

### My CA files
From the server machine, logged in as root
Generate CA private key ca-priv.key
`openssl genrsa -des3 -out /etc/ssl/private/ca-priv.key 2048`

Set permissions on CA private key ca-priv.key so there is no group or world access
`chmod 600 /etc/ssl/private/ca-priv.key`

Generate CA certificate ca-priv.crt using CA private key
From the server
`sudo openssl req -new -x509 -days 3650 
-subj '/C={MyCountry}/ST={My State}/L={My Location}/O={My Org}/CN={my CA name}' 
-key /etc/ssl/private/ca-priv.key 
-out /usr/local/share/ca-certificates/ca-cert.crt`

Add certificate to the server trusted roots
`update-ca-certificates`

### Server Files
From the server machine, logged in as root
Generate server private key server-priv.key
`openssl genrsa -des3 -out /etc/ssl/private/server-priv.key 2048`

Remove the passphrase from server private key server-priv.key
`openssl rsa -in /etc/ssl/private/server-priv.key -out /etc/ssl/private/server-priv.key`

Set permissions on server private key server-priv.key so world has no access
`chmod 640 /etc/ssl/private/server-priv.key`

Generate server certificate request server-cert-req.csr
openssl req -new -nodes -key /etc/ssl/private/server-priv.key -days 3650 -out /tmp/server-cert-req.csr -subj '/C={My Country}/ST={My State}/L={My Location}/O={My Org}/CN={my server name}'

Generate (sign) the server certificate request by 
`openssl x509 -days 3650 
-req -in /tmp/server-cert-req.csr 
-CA /usr/local/share/ca-certificates/ca-cert.crt 
-CAkey /etc/ssl/private/ca-priv.key -CAcreateserial 
-out /usr/local/share/ca-certificates/server-cert.crt`

Generate the server certificate that includes the full chain back to the root signing CA 
`cat /usr/local/share/ca-certificates/server-cert.crt /usr/local/share/ca-certificates/ca-cert.crt > /usr/local/share/ca-certificates/server-cert-fullchain.crt`
Because there is no intermediary between the server and the root, this step is not really required by client

Log into the postgreSQL sever, create the group sslcert and add user john to the group
`su - postgres psql`
`CREATE GROUP sslcert;`
`ALTER GROUP sslcert ADD USER john;`

Modify PostgreSQL /etc/postgresql/9.6/main/pg_ident.conf:
`# MAPNAME       SYSTEM-USERNAME         PG-USERNAME`
`mymap         johnsmith              john`

Modify PostgreSQL /etc/postgresql/9.6/main/postgresql.conf for SSL
`ssl = on`
`ssl_ciphers = 'HIGH:MEDIUM:+3DES:!aNULL'`
`ssl_prefer_server_ciphers = on`
`#ssl_ecdh_curve = 'prime256v1'` 
`ssl_cert_file = '/etc/ssl/certs/server-cert.pem'`
`ssl_key_file = '/etc/ssl/private/server-priv.key'`
`ssl_ca_file = '/etc/ssl/certs/ca-cert.pem'`
`#ssl_crl_file = ''`

Modify PostgreSQL /etc/postgresql/9.6/main/pg_hba.conf:
`hostssl all +sslcertusers all cert clientcert=1 map=mymap`

### User Files
From the server machine, logged in as john
Generate my private key file johnsmith-priv.key
`openssl genrsa -des3 -out ~/johnsmith-priv.key 2048`

Remove the passphrase from my user private key johnsmith-priv.key
`openssl rsa -in ~/johnsmith-priv.key -out ~/johnsmith-priv.key`

Generate my user certificate request johnsmith-cert-req.csr
openssl req -new -nodes -key ~/johnsmith-priv.key -days 3650 -out ~/johnsmith-cert-req.csr -subj '/C={My Country}/ST={My State}/L={My Location}/O={My Org}/CN=johnsmith'

Request my CA to sign my user certificate request
sudo openssl x509 -days 3650 -req -in ~/johnsmith-cert-req.csr -CA /etc/ssl/certs/ca-cert.pem -CAkey /etc/ssl/private/ca-priv.key -out ~/johnsmith-cert.crt -CAcreateserial

From the client machine, logged in as johnsmith

create the .postgresql directory
`mkdir ~/.postgresql`

copy these certificate and private key files from the server to the client
These filenames are the recommended names according to the PostgreSQL docs
`scp john@server.com:/user/local/share/ca-certificates/ca-cert.crt ~/.postgresql/root.crt`
`scp john@server.com:/user/local/share/ca-certificates/server-cert.crt ~/.postgresql/postgresql.crt`
`scp john@server.com:/johnsmith-pri.key ~/.postgresql/postgresql.key`

set the permissions on postgresql.key so no world access
`chmod 640 ~/.postgresql/postgresql.key`

### Start DBeaver
Create a database connection to server.com with the following settings
`user: john`
`database: postgres`
`ssl: checked`
`root: ~/.postgresql/root.crt`
`ssl certificate:  ~/.postgresql/postgresql.crt`
`ssl certificate key:  ~/.postgresql/postgresql.key`
`ssl mode: verify-ca`

I get the following error:
`Could not read SSL key file /Users/johnsmith/.postgresql/postgresql.key.  java.io.IOException: extra data given to DerValue constructor`

I can confirm the SSL configuration on the server is correct because I can connect to it from application PGAdmin4.

Any ideas how can I resolve this error?

**Comment by msluser** [comment]
Ok, according to Russell Gray at his [blog post](https://basildoncoder.com/blog/postgresql-jdbc-client-certificates.html)

> In the org.postgresql.ssl.jdbc4 package we find the LibPQFactory class, which handles everything we need - and, in what I personally consider a bonus, works with .crt files directly rather than mucking about with keystores and the perenially stupid keytool.
> 
> By default, this class looks in ~/.postgresql/ for the files it needs to work, though you can override the location. You need a root.crt containing the CA for the server certificate, plus your client certificate (postgresql.crt) and private key (postgresql.pk8).
> 

...

> Although psql has no trouble, the JDBC driver has format restrictions. The private key must be PKCS8 and stored in DER format, whereas the certificate is fine in PEM format (because of course it is). If your key is in PEM format (i.e. starts with something like -----BEGIN PRIVATE KEY-----) you can convert it with openssl:

`openssl pkcs8 -topk8 -inform PEM -outform DER -in postgresql.key -out postgresql`

So next I converted my private key from PEM to DER format using the command above.

Next I had to remove any group or world permissions on my private key file

`chmod 600 ~./postgresql/postgresql.pk8`

Now for the JDBC driver to utilize my private key I had to provide the key passphrase by setting driver property:
`sslpassword=mysecret`

Now I am able to progressed further, however I am ending with this error.

![image](https://user-images.githubusercontent.com/20219053/27844660-c1d592fa-60e1-11e7-8cbd-b39295f3d750.png)
The tail lines from the log file is
```
2017-07-04 17:53:14.524 - org.jkiss.dbeaver.model.exec.DBCConnectException: SSL error: org.eclipse.swt.SWTException: Invalid thread access
org.jkiss.dbeaver.model.exec.DBCConnectException: SSL error: org.eclipse.swt.SWTException: Invalid thread access
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:190)
	at org.jkiss.dbeaver.ext.postgresql.model.PostgreDataSource.openConnection(PostgreDataSource.java:348)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:72)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.initializeMainContext(JDBCDataSource.java:115)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:109)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:98)
	at org.jkiss.dbeaver.ext.postgresql.model.PostgreDataSource.<init>(PostgreDataSource.java:73)
	at org.jkiss.dbeaver.ext.postgresql.PostgreDataSourceProvider.openDataSource(PostgreDataSourceProvider.java:89)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:682)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:70)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionWizard$ConnectionTester.run(ConnectionWizard.java:213)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:95)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: org.postgresql.util.PSQLException: SSL error: org.eclipse.swt.SWTException: Invalid thread access
	at org.postgresql.ssl.MakeSSL.convert(MakeSSL.java:67)
	at org.postgresql.core.v3.ConnectionFactoryImpl.enableSSL(ConnectionFactoryImpl.java:359)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:148)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:194)
	at org.postgresql.Driver.makeConnection(Driver.java:450)
	at org.postgresql.Driver.connect(Driver.java:252)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:176)
	... 12 more
Caused by: javax.net.ssl.SSLException: org.eclipse.swt.SWTException: Invalid thread access
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:208)
	at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1949)
	at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1906)
	at sun.security.ssl.SSLSocketImpl.handleException(SSLSocketImpl.java:1889)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1410)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
	at org.postgresql.ssl.MakeSSL.convert(MakeSSL.java:62)
	... 19 more
Caused by: org.eclipse.swt.SWTException: Invalid thread access
	at org.eclipse.swt.SWT.error(SWT.java:4533)
	at org.eclipse.swt.SWT.error(SWT.java:4448)
	at org.eclipse.swt.SWT.error(SWT.java:4419)
	at org.eclipse.swt.widgets.Display.error(Display.java:1089)
	at org.eclipse.swt.widgets.Display.checkDevice(Display.java:632)
	at org.eclipse.swt.widgets.Display.getActiveShell(Display.java:1252)
	at org.jkiss.dbeaver.core.DBeaverUI.getActiveWorkbenchShell(DBeaverUI.java:189)
	at org.jkiss.dbeaver.core.DBeaverUI.promptUserCredentials(DBeaverUI.java:401)
	at org.jkiss.dbeaver.runtime.net.DefaultCallbackHandler.handle(DefaultCallbackHandler.java:40)
	at org.postgresql.ssl.jdbc4.LazyKeyManager.getPrivateKey(LazyKeyManager.java:207)
	at sun.security.ssl.AbstractKeyManagerWrapper.getPrivateKey(SSLContextImpl.java:1240)
	at sun.security.ssl.ClientHandshaker.serverHelloDone(ClientHandshaker.java:824)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:348)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:1026)
	at sun.security.ssl.Handshaker.process_record(Handshaker.java:961)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1062)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
```

Interestingly, I can verify this JDBC driver and settings work with the java application [Eclipse BIRT](http://download.eclipse.org/birt/downloads/).
So I don't know why this is failing with DBeaver.

Now, if I remove the passphrase encryption from my private key using

`openssl pkcs8 -topk8 -inform PEM -outform DER -in postgresql.key -out postgresql -nocrypt`

I can connect!

![image](https://user-images.githubusercontent.com/20219053/27844679-09315a4e-60e2-11e7-8625-ebcda45e2bf3.png)

Why is a thread error being generated when I use a passphrase on my private key?

**Comment by serge-rider** [comment]
This (invalid thread access error) is a bug in DBeaver. Thanks for report.

**Comment by radcapitalist** [comment]
I followed the instructions here to convert my key to PK8 format, and I set the passphrase in the **sslpassword** driver property in the connection.  However, when I try to connect, I get prompted for the password.  I enter the password the first time, the dialog goes away and comes back.  I enter the password a second time.  Same thing.  If I enter the password 4 times, then the password dialog stops popping up and the connection is successful.  DBeaver version is 7.3.4.
UPDATE: I am prompted 5 times, not 4.  I have **sslmode** set to **verify-ca**, but **require** and **prefer** show the same behavior.
